base/broadcast.jl

```diff
@@ -102,15 +94,14 @@ behaves as an `N`-dimensional array for broadcasting. Specifically, `DefaultArra
 used for any
 AbstractArray type that hasn't defined a specialized style, and in the absence of
 overrides from other `broadcast` arguments the resulting output type is `Array`.
-When there are multiple inputs to `broadcast`, `DefaultArrayStyle` "wins" over [`Broadcast.Scalar`](@ref)
-but "loses" to any other [`Broadcast.ArrayStyle`](@ref).
+When there are multiple inputs to `broadcast`, `DefaultArrayStyle` "loses" to any other [`Broadcast.ArrayStyle`](@ref).
 """
 struct DefaultArrayStyle{N} <: AbstractArrayStyle{N} end
 (::Type{<:DefaultArrayStyle})(::Val{N}) where N = DefaultArrayStyle{N}()
 const DefaultVectorStyle = DefaultArrayStyle{1}
 const DefaultMatrixStyle = DefaultArrayStyle{2}
 BroadcastStyle(::Type{<:AbstractArray{T,N}}) where {T,N} = DefaultArrayStyle{N}()
-BroadcastStyle(::Type{<:Ref}) = DefaultArrayStyle{0}()
+BroadcastStyle(::Type{<:Union{Base.RefValue,Number}}) = DefaultArrayStyle{0}()
```

Comment:
Why is that a good thing (I don't actually know what this return value means)? For context, `RefArray` is a pointer to precisely one element of an array. It also changes `Ptr`, if that matters?

Reply:
All arguments' `BroadcastStyle`s are combined with a promote-like mechanism in order to select a broadcasting implementation.  For the most part, it's largely used as a kludge around the lack of a "at least one vararg is of type `T`" dispatch, but it's more complicated to deal with cases that would otherwise be ambiguities.

Yes, `Ptr` has neither axes or indexing defined, so it cannot be — by itself — considered like a 0-dimensional array.  I don't think `RefArray` is common enough to add special broadcasting support for it.

Before:

```
julia> identity.(Ref([1,2,3],1))
0-dimensional Array{Int64,0}:
1
```

After:

```
julia> identity.(Ref([1,2,3],1))
┌ Warning: broadcast will default to iterating over its arguments in the future. Wrap arguments of
│ type `x::Base.RefArray{Int64,Array{Int64,1},Nothing}` with `Ref(x)` to ensure they broadcast as "scalar" elements.
│   caller = top-level scope
└ @ Core :0
Base.RefArray{Int64,Array{Int64,1},Nothing}([1, 2, 3], 1, nothing)
```

In the future, it will be an error because:

```
julia> collect(Ref([1,2,3],1))
ERROR: MethodError: no method matching length(::Base.RefArray{Int64,Array{Int64,1},Nothing})
```

Reply:
> I don't think RefArray is common enough to add special broadcasting support for it.

However, `RefValue` is an implementation detail, and should never be mentioned explicitly either: `RefArray` is simply an alternative implementation of it. We know that `length(::Ref) == 1` forall Ref subtypes, we've just hesitated to add that definition to the code until needed to keep the API small.

Reply:
This is largely a historical accident due to how I ended up here… and trying to work around `Ptr`.  As it stands, I can easily remove all uses of `RefValue`.  I just pushed a change.

New after:

```
julia> identity.(Ref([1,2,3],1))
1
```